// react-billboardjs/lib/shapes.js
'use strict';

exports.__esModule = true;
exports.ZOOM_SHAPE = exports.TRANSITION_SHAPE = exports.TOOLTIP_SHAPE = exports.TITLE_SHAPE = exports.SVG_SHAPE = exports.SUBCHART_SHAPE = exports.SPLINE_SHAPE = exports.SIZE_SHAPE = exports.RESIZE_SHAPE = exports.REGION_SHAPE = exports.POINT_SHAPE = exports.PIE_SHAPE = exports.PADDING_SHAPE = exports.LINE_SHAPE = exports.LEGEND_SHAPE = exports.INTERACTION_SHAPE = exports.GRID_SHAPE = exports.LINES_SHAPE = exports.GAUGE_SHAPE = exports.DONUT_SHAPE = exports.DATA_SHAPE = exports.ORDER_SHAPE = exports.COLOR_SHAPE = exports.BAR_SHAPE = exports.AXIS_SHAPE = exports.AXIS_TICK_SHAPE = exports.AXIS_Y_PADDING_SHAPE = exports.AXIS_TYPE_SHAPE = exports.LABEL_SHAPE = exports.AREA_SHAPE = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var AREA_SHAPE = exports.AREA_SHAPE = _propTypes2.default.shape({
  above: _propTypes2.default.bool,
  zerobased: _propTypes2.default.bool
}); // external dependencies
var LABEL_SHAPE = exports.LABEL_SHAPE = _propTypes2.default.oneOfType([_propTypes2.default.shape({
  position: _propTypes2.default.oneOf(['inner-bottom', 'inner-center', 'inner-left', 'inner-middle', 'inner-right', 'inner-top', 'outer-bottom', 'outer-center', 'outer-left', 'outer-middle', 'outer-right', 'outer-top']),
  text: _propTypes2.default.string
}), _propTypes2.default.string]);

var AXIS_TYPE_SHAPE = exports.AXIS_TYPE_SHAPE = _propTypes2.default.oneOf(['category', 'indexed', 'timeseries']);

var AXIS_Y_PADDING_SHAPE = exports.AXIS_Y_PADDING_SHAPE = _propTypes2.default.shape({
  bottom: _propTypes2.default.number,
  top: _propTypes2.default.number
});

var AXIS_TICK_SHAPE = exports.AXIS_TICK_SHAPE = _propTypes2.default.shape({
  count: _propTypes2.default.number,
  format: _propTypes2.default.func,
  values: _propTypes2.default.arrayOf(_propTypes2.default.number)
});

var AXIS_SHAPE = exports.AXIS_SHAPE = _propTypes2.default.shape({
  rotated: _propTypes2.default.bool,
  x: _propTypes2.default.shape({
    categories: _propTypes2.default.arrayOf(_propTypes2.default.string),
    extent: _propTypes2.default.arrayOf(_propTypes2.default.arrayOf(_propTypes2.default.number)),
    height: _propTypes2.default.number,
    label: LABEL_SHAPE,
    localtime: _propTypes2.default.bool,
    max: _propTypes2.default.number,
    min: _propTypes2.default.number,
    padding: _propTypes2.default.shape({
      left: _propTypes2.default.number,
      right: _propTypes2.default.number
    }),
    show: _propTypes2.default.bool,
    tick: _propTypes2.default.shape({
      centered: _propTypes2.default.bool,
      count: _propTypes2.default.number,
      culling: _propTypes2.default.bool,
      cullingMax: _propTypes2.default.number,
      fit: _propTypes2.default.bool,
      format: _propTypes2.default.func,
      multiline: _propTypes2.default.bool,
      outer: _propTypes2.default.bool,
      rotate: _propTypes2.default.number,
      tooltip: _propTypes2.default.bool,
      values: _propTypes2.default.arrayOf(_propTypes2.default.number),
      width: _propTypes2.default.number
    }),
    type: AXIS_TYPE_SHAPE
  }),
  y: _propTypes2.default.shape({
    center: _propTypes2.default.number,
    default: _propTypes2.default.arrayOf(_propTypes2.default.number),
    format: _propTypes2.default.func,
    inner: _propTypes2.default.bool,
    inverted: _propTypes2.default.bool,
    label: LABEL_SHAPE,
    max: _propTypes2.default.number,
    min: _propTypes2.default.number,
    padding: AXIS_Y_PADDING_SHAPE,
    show: _propTypes2.default.bool,
    tick: AXIS_TICK_SHAPE,
    type: AXIS_TYPE_SHAPE
  }),
  y2: _propTypes2.default.shape({
    center: _propTypes2.default.number,
    default: _propTypes2.default.arrayOf(_propTypes2.default.number),
    inner: _propTypes2.default.bool,
    inverted: _propTypes2.default.bool,
    label: LABEL_SHAPE,
    max: _propTypes2.default.number,
    min: _propTypes2.default.number,
    padding: AXIS_Y_PADDING_SHAPE,
    show: _propTypes2.default.bool,
    tick: AXIS_TICK_SHAPE
  })
});

var BAR_SHAPE = exports.BAR_SHAPE = _propTypes2.default.shape({
  padding: _propTypes2.default.number,
  width: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.shape({
    max: _propTypes2.default.number,
    ratio: _propTypes2.default.number
  })]),
  zerobased: _propTypes2.default.bool
});

var COLOR_SHAPE = exports.COLOR_SHAPE = _propTypes2.default.shape({
  pattern: _propTypes2.default.arrayOf(_propTypes2.default.string),
  threshold: _propTypes2.default.shape({
    max: _propTypes2.default.number,
    unit: _propTypes2.default.string,
    value: _propTypes2.default.string
  })
});

var ORDER_SHAPE = exports.ORDER_SHAPE = _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.oneOf(['asc', 'desc', '', null])]);

var DATA_SHAPE = exports.DATA_SHAPE = _propTypes2.default.shape({
  axes: _propTypes2.default.object,
  classes: _propTypes2.default.object,
  color: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.shape({
    pattern: _propTypes2.default.arrayOf(_propTypes2.default.string),
    tiles: _propTypes2.default.func
  })]),
  colors: _propTypes2.default.object,
  columns: _propTypes2.default.arrayOf(_propTypes2.default.arrayOf(_propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.number), _propTypes2.default.number, _propTypes2.default.shape({
    height: _propTypes2.default.number,
    low: _propTypes2.default.number,
    mid: _propTypes2.default.number
  }), _propTypes2.default.string]))),
  empty: _propTypes2.default.shape({
    label: LABEL_SHAPE
  }),
  groups: _propTypes2.default.arrayOf(_propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.string), _propTypes2.default.string])),
  hide: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.arrayOf(_propTypes2.default.string)]),
  json: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.object), _propTypes2.default.object]),
  keys: _propTypes2.default.shape({
    value: _propTypes2.default.arrayOf(_propTypes2.default.string),
    x: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.string), _propTypes2.default.string]),
    y: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.string), _propTypes2.default.string]),
    y2: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.string), _propTypes2.default.string])
  }),
  labels: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.shape({
    format: _propTypes2.default.func,
    position: _propTypes2.default.shape({
      x: _propTypes2.default.number,
      y: _propTypes2.default.number
    })
  })]),
  line: _propTypes2.default.shape({
    classes: _propTypes2.default.arrayOf(_propTypes2.default.string)
  }),
  mimeType: _propTypes2.default.string,
  names: _propTypes2.default.object,
  onclick: _propTypes2.default.func,
  onmax: _propTypes2.default.func,
  onmin: _propTypes2.default.func,
  onout: _propTypes2.default.func,
  onover: _propTypes2.default.func,
  onselected: _propTypes2.default.func,
  onunselected: _propTypes2.default.func,
  order: ORDER_SHAPE,
  point: _propTypes2.default.shape({
    pattern: _propTypes2.default.arrayOf(_propTypes2.default.string),
    type: _propTypes2.default.oneOf(['circle', 'rectangle'])
  }),
  regions: _propTypes2.default.object,
  rows: _propTypes2.default.arrayOf(_propTypes2.default.arrayOf(_propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string]))),
  selection: _propTypes2.default.shape({
    draggable: _propTypes2.default.bool,
    enabled: _propTypes2.default.bool,
    isselectable: _propTypes2.default.bool,
    multiple: _propTypes2.default.bool
  }),
  type: _propTypes2.default.oneOf(['area', 'area-line-range', 'area-range', 'area-spline', 'area-step', 'bar', 'bubble', 'donut', 'gauge', 'line', 'pie', 'scatter', 'spline']),
  types: _propTypes2.default.object,
  url: _propTypes2.default.string,
  x: _propTypes2.default.string,
  xFormat: _propTypes2.default.string,
  xLocaltime: _propTypes2.default.bool,
  xs: _propTypes2.default.object,
  xSort: _propTypes2.default.bool
});

var DONUT_SHAPE = exports.DONUT_SHAPE = _propTypes2.default.shape({
  label: _propTypes2.default.shape({
    format: _propTypes2.default.func,
    radio: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.number]),
    show: _propTypes2.default.bool,
    threshold: _propTypes2.default.number
  }),
  expand: _propTypes2.default.bool,
  padAngle: _propTypes2.default.number,
  title: _propTypes2.default.string,
  width: _propTypes2.default.number
});

var GAUGE_SHAPE = exports.GAUGE_SHAPE = _propTypes2.default.shape({
  expand: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.shape({
    duration: _propTypes2.default.number
  })]),
  fullCircle: _propTypes2.default.bool,
  label: _propTypes2.default.shape({
    extents: _propTypes2.default.func,
    format: _propTypes2.default.func,
    show: _propTypes2.default.bool
  }),
  max: _propTypes2.default.number,
  startingAngle: _propTypes2.default.number,
  units: _propTypes2.default.string,
  width: _propTypes2.default.number
});

var LINES_SHAPE = exports.LINES_SHAPE = _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.shape({
  class: _propTypes2.default.string,
  position: _propTypes2.default.string,
  text: _propTypes2.default.string,
  value: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.string])
})), _propTypes2.default.bool]);

var GRID_SHAPE = exports.GRID_SHAPE = _propTypes2.default.shape({
  focus: _propTypes2.default.shape({
    show: _propTypes2.default.bool
  }),
  lines: _propTypes2.default.shape({
    front: _propTypes2.default.bool
  }),
  x: _propTypes2.default.shape({
    lines: LINES_SHAPE,
    show: _propTypes2.default.bool
  }),
  y: _propTypes2.default.shape({
    lines: LINES_SHAPE,
    show: _propTypes2.default.bool,
    ticks: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.number])
  })
});

var INTERACTION_SHAPE = exports.INTERACTION_SHAPE = _propTypes2.default.shape({
  brighten: _propTypes2.default.bool,
  enabled: _propTypes2.default.bool,
  inputType: _propTypes2.default.shape({
    mouse: _propTypes2.default.bool,
    touch: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.shape({
      preventDefault: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.number])
    })])
  })
});

var LEGEND_SHAPE = exports.LEGEND_SHAPE = _propTypes2.default.shape({
  contents: _propTypes2.default.shape({
    bindto: _propTypes2.default.oneOfType([_propTypes2.default.object, _propTypes2.default.string]),
    template: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.string])
  }),
  equally: _propTypes2.default.bool,
  hide: _propTypes2.default.bool,
  inset: _propTypes2.default.shape({
    anchor: _propTypes2.default.oneOf(['bottom-left', 'bottom-right', 'top-left', 'top-right']),
    step: _propTypes2.default.number,
    x: _propTypes2.default.number,
    y: _propTypes2.default.number
  }),
  item: _propTypes2.default.shape({
    onclick: _propTypes2.default.func,
    onout: _propTypes2.default.func,
    onover: _propTypes2.default.func,
    tile: _propTypes2.default.shape({
      height: _propTypes2.default.number,
      width: _propTypes2.default.number
    })
  }),
  padding: _propTypes2.default.number,
  position: _propTypes2.default.oneOf(['bottom', 'right', 'inset']),
  show: _propTypes2.default.bool,
  usePoint: _propTypes2.default.bool
});

var LINE_SHAPE = exports.LINE_SHAPE = _propTypes2.default.shape({
  connectNull: _propTypes2.default.bool,
  step: _propTypes2.default.shape({
    type: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.oneOf(['step', 'step-after', 'step-before'])])
  })
});

var PADDING_SHAPE = exports.PADDING_SHAPE = _propTypes2.default.shape({
  bottom: _propTypes2.default.number,
  left: _propTypes2.default.number,
  right: _propTypes2.default.number,
  top: _propTypes2.default.number
});

var PIE_SHAPE = exports.PIE_SHAPE = _propTypes2.default.shape({
  expand: _propTypes2.default.bool,
  innerRadius: _propTypes2.default.number,
  label: _propTypes2.default.shape({
    format: _propTypes2.default.func,
    ratio: _propTypes2.default.number,
    show: _propTypes2.default.bool,
    threshold: _propTypes2.default.number
  }),
  padAngle: _propTypes2.default.number,
  padding: _propTypes2.default.number
});

var POINT_SHAPE = exports.POINT_SHAPE = _propTypes2.default.shape({
  focus: _propTypes2.default.shape({
    expand: _propTypes2.default.shape({
      enabled: _propTypes2.default.bool,
      r: _propTypes2.default.bool
    })
  }),
  r: _propTypes2.default.number,
  select: _propTypes2.default.shape({
    r: _propTypes2.default.number
  }),
  show: _propTypes2.default.bool
});

var REGION_SHAPE = exports.REGION_SHAPE = _propTypes2.default.shape({
  axis: _propTypes2.default.string,
  class: _propTypes2.default.string,
  end: _propTypes2.default.number,
  start: _propTypes2.default.number
});

var RESIZE_SHAPE = exports.RESIZE_SHAPE = _propTypes2.default.shape({
  auto: _propTypes2.default.bool
});

var SIZE_SHAPE = exports.SIZE_SHAPE = _propTypes2.default.shape({
  height: _propTypes2.default.number,
  width: _propTypes2.default.number
});

var SPLINE_SHAPE = exports.SPLINE_SHAPE = _propTypes2.default.shape({
  interpolation: _propTypes2.default.shape({
    type: _propTypes2.default.string
  })
});

var SUBCHART_SHAPE = exports.SUBCHART_SHAPE = _propTypes2.default.shape({
  onbrush: _propTypes2.default.oneOfType([_propTypes2.default.bool, _propTypes2.default.func]),
  size: _propTypes2.default.shape({
    height: _propTypes2.default.number
  }),
  show: _propTypes2.default.bool
});

var SVG_SHAPE = exports.SVG_SHAPE = _propTypes2.default.shape({
  className: _propTypes2.default.string
});

var TITLE_SHAPE = exports.TITLE_SHAPE = _propTypes2.default.shape({
  padding: PADDING_SHAPE,
  position: _propTypes2.default.oneOf(['bottom-center', 'bottom-left', 'bottom-right', 'top-center', 'top-left', 'top-right']),
  text: _propTypes2.default.string
});

var TOOLTIP_SHAPE = exports.TOOLTIP_SHAPE = _propTypes2.default.shape({
  contents: _propTypes2.default.func,
  format: _propTypes2.default.shape({
    name: _propTypes2.default.func,
    title: _propTypes2.default.func,
    value: _propTypes2.default.func
  }),
  grouped: _propTypes2.default.bool,
  linked: _propTypes2.default.bool,
  order: ORDER_SHAPE,
  position: _propTypes2.default.func,
  show: _propTypes2.default.bool
});

var TRANSITION_SHAPE = exports.TRANSITION_SHAPE = _propTypes2.default.shape({
  duration: _propTypes2.default.number
});

var ZOOM_SHAPE = exports.ZOOM_SHAPE = _propTypes2.default.shape({
  enabled: _propTypes2.default.bool,
  extent: _propTypes2.default.arrayOf(_propTypes2.default.number),
  onzoom: _propTypes2.default.func,
  onzoomend: _propTypes2.default.func,
  onzoomstart: _propTypes2.default.func,
  rescale: _propTypes2.default.bool
});
